src/python-samplerate.cpp

```diff
@@ -0,0 +1,570 @@
+
+  // get the data as a numpy array
+  auto input = cb->callback();
+
+  // accessors for the arrays
+  py::buffer_info inbuf = input.request();
+
+  // end of stream is signaled by a None, which is cast to a ndarray with ndim
+  // == 0
+  if (inbuf.ndim == 0) return 0;
+
+  // set the number of channels
+  int channels = 1;
+  if (inbuf.ndim == 2)
+    channels = inbuf.shape[1];
+  else if (inbuf.ndim > 2)
+    throw std::domain_error("Input array should have at most 2 dimensions");
+
+  if (channels != cb_channels || channels == 0)
+    throw std::domain_error("Invalid number of channels in input data.");
+
+  *data = static_cast<float *>(inbuf.ptr);
+
+  return (long)inbuf.shape[0];
+}
+
+py::array_t<float, py::array::c_style> resample(
+    const py::array_t<float, py::array::c_style | py::array::forcecast> &input,
+    double sr_ratio, const py::object &converter_type, bool verbose) {
+  // input array has shape (n_samples, n_channels)
+  int converter_type_int = get_converter_type(converter_type);
+
+  // accessors for the arrays
+  py::buffer_info inbuf = input.request();
+
+  // set the number of channels
+  int channels = 1;
+  if (inbuf.ndim == 2)
+    channels = inbuf.shape[1];
+  else if (inbuf.ndim > 2)
+    throw std::domain_error("Input array should have at most 2 dimensions");
+
+  if (channels == 0)
+    throw std::domain_error("Invalid number of channels (0) in input data.");
+
+  const size_t new_size = size_t(std::ceil(inbuf.shape[0] * sr_ratio));
+
+  // allocate output array
+  std::vector<size_t> out_shape{new_size};
+  if (inbuf.ndim == 2) out_shape.push_back(size_t(channels));
+  auto output = py::array_t<float, py::array::c_style>(out_shape);
+  py::buffer_info outbuf = output.request();
+
+  // libsamplerate struct
+  SRC_DATA src_data = {
+      static_cast<float *>(inbuf.ptr),   // data_in
+      static_cast<float *>(outbuf.ptr),  // data_out
+      inbuf.shape[0],                    // input_frames
+      long(new_size),                    // output_frames
+      0,        // input_frames_used, filled by libsamplerate
+      0,        // output_frames_gen, filled by libsamplerate
+      0,        // end_of_input, not used by src_simple ?
+      sr_ratio  // src_ratio, sampling rate conversion ratio
+  };
+
+  int ret_code = src_simple(&src_data, converter_type_int, channels);
+
+  error_handler(ret_code);
+
+  // create a shorter view of the array
+  if ((size_t)src_data.output_frames_gen < new_size) {
+    out_shape[0] = src_data.output_frames_gen;
+    output = py::array_t<float, py::array::c_style>(
+        out_shape, static_cast<float *>(outbuf.ptr));
+  }
+
+  if (verbose) {
+    std::cout << "samplerate info:\n"
+              << src_data.input_frames_used << " input frames used\n"
+              << src_data.output_frames_gen << " output frames generated\n";
+  }
+
+  return output;
+}
+
+}  // namespace samplerate
+
+namespace sr = samplerate;
+
+PYBIND11_MODULE(samplerate, m) {
```

Comment:
Subpackages `samplerate.converters` and `samplerate.exceptions` should stay to maintain compatibility. `samplerate.lowlevel` can just go, I don't think anybody should depend on it.

Reply:
In the original package, the converters are accessible both at the package level and `samplerate.converters` sub-package level.
Do you know if there is a nice way to do that with pybind11 ?